runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPipelineTranslator.java

```diff
@@ -222,7 +282,117 @@ boolean canTranslate(T transform, FlinkStreamingTranslationContext context) {
     @Override
     public Map<PValue, ReplacementOutput> mapOutputs(
         Map<TupleTag<?>, PValue> outputs, WriteFilesResult<DestinationT> newOutput) {
-      return Collections.emptyMap();
+      return ReplacementOutputs.tagged(outputs, newOutput);
+    }
+  }
+
+  /**
+   * Flink has a known problem of unevenly assigning keys to key groups (and then workers) for cases
+   * that number of keys is not >> key groups. This is typical scenario when writing files, where
```

Comment:
I wouldn't say this is a Flink problem, it just comes with the nature of hashing algorithms (Murmur hash in this case) when the number of keys is not much bigger than the number of partitions.

Reply:
True. Does this sounds better?
```
Flink assigns elements to parallel operators (workers) based their key group. Key group is determined based on a Murmur hash of element's key. This will have skew distribution in case when number of keys is not >> key groups, which is typical scenario when writing files, where one do not want to end up with too many small files.
```

Reply:
Sounds good!